File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/entity/Entity.java
/**
 * 
 */
package com.abubusoft.xenon.entity;


import com.abubusoft.xenon.mesh.Mesh;

/**
 * @author Francesco Benincasa
 * 
 */
public class Entity<E extends Mesh> extends BaseEntity {

	private static final long serialVersionUID = 3212120492925326047L;

	public Entity(E shapeValue) {
		this();
		mesh = shapeValue;
	}

	/**
	 * Costruttore di default
	 */
	public Entity() {
		super();
		visible = true;
		collidable = true;
	}

	/**
	 * indicase se l'entità è visibile
	 */
	public boolean visible;

	/**
	 * indica se l'entità è soggetta a collisioni
	 */
	public boolean collidable;

	/**
	 * mesh associato all'entity
	 */
	public E mesh;

	/**
	 * Raggio del cerchio di contenimento
	 * 
	 * @return
	 */
	public float getBoundingRadius() {
		return mesh.boundingSphereRadius;
	}

	/* (non-Javadoc)
	 * @see com.abubusoft.xenon.core.util.Copy#copy()
	 */
	@SuppressWarnings("unchecked")
	@Override
	public Entity<E> copy() {
		Entity<E> copy=null;
		try {
			// cloniamo
			copy = getClass().newInstance();
			copyInto(copy);						
		} catch (InstantiationException e) {
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		}
		return copy;
	}

	/**
	 * Effettua la copia nell'oggetto destinazione
	 * 
	 * @param destination
	 */
	public void copyInto(Entity<E> destination) {
		super.copyInto(destination);
		destination.collidable=collidable;
		destination.mesh=mesh;
		destination.visible=visible;
	}

}

File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/entity/Entity.kt
/**
 *
 */
package com.abubusoft.xenon.entity

import com.abubusoft.xenon.mesh.Mesh

/**
 * @author Francesco Benincasa
 */
open class Entity<E : Mesh?>
/**
 * Costruttore di default
 */() : BaseEntity() {
    constructor(shapeValue: E) : this() {
        mesh = shapeValue
    }

    /**
     * indicase se l'entità è visibile
     */
    var visible = true

    /**
     * indica se l'entità è soggetta a collisioni
     */
    var collidable = true

    /**
     * mesh associato all'entity
     */
    var mesh: E? = null

    /**
     * Raggio del cerchio di contenimento
     *
     * @return
     */
    open val boundingRadius: Float
        get() = mesh!!.boundingSphereRadius

    /* (non-Javadoc)
	 * @see com.abubusoft.xenon.core.util.Copy#copy()
	 */
    override fun copy(): Entity<E?>? {
        var copy: Entity<E?>? = null
        try {
            // cloniamo
            copy = javaClass.newInstance()
            copyInto(copy)
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return copy
    }

    /**
     * Effettua la copia nell'oggetto destinazione
     *
     * @param destination
     */
    fun copyInto(destination: Entity<E?>?) {
        super.copyInto(destination)
        destination!!.collidable = collidable
        destination.mesh = mesh
        destination.visible = visible
    }

    companion object {
        private const val serialVersionUID = 3212120492925326047L
    }
}